### PR TITLE
!str split Framing into javadsl and scaladsl

### PR DESCRIPTION
**Pull request by ktoso**

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+   * Creates a Flow that handles decoding a stream of unstructured byte chunks into a stream of frames where the
+   * incoming chunk stream uses a specific byte-sequence to mark frame boundaries.
+   *
+   * The decoded frames will not include the separator sequence.
+   *
+   * If there are buffered bytes (an incomplete frame) when the input stream finishes and ''allowTruncation'' is set to
+   * false then this Flow will fail the stream reporting a truncated frame.
+   *
+   * @param delimiter The byte sequence to be treated as the end of the frame.
+   * @param allowTruncation If turned on, then when the last frame being decoded contains no valid delimiter this Flow
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: the opposite. If truncation is allowed, then it does not fail the stream if its, well, truncated :)

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: Also, maybe say, "set to ALLOW"

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+   * @param fieldOffset The offset of the field from the beginning of the frame in bytes
+   * @param maximumFrameLength The maximum length of allowed frames while decoding. If the maximum length is exceeded
+   *                           this Flow will fail the stream. This length *includes* the header (i.e the offset and
+   *                           the length of the size field)
+   * @param byteOrder The ''ByteOrder'' to be used when decoding the field
+   */
+  def lengthField(fieldLength: Int,
+                  fieldOffset: Int,
+                  maximumFrameLength: Int,
+                  byteOrder: ByteOrder = ByteOrder.LITTLE_ENDIAN): Flow[ByteString, ByteString, NotUsed] =
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: no default arguments in Java DSL please (they create weird method names)

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+   * incoming frames have a field that encodes their length.
+   *
+   * If the input stream finishes before the last frame has been fully decoded this Flow will fail the stream reporting
+   * a truncated frame.
+   *
+   * @param fieldLength The length of the "size" field in bytes
+   * @param fieldOffset The offset of the field from the beginning of the frame in bytes
+   * @param maximumFrameLength The maximum length of allowed frames while decoding. If the maximum length is exceeded
+   *                           this Flow will fail the stream. This length *includes* the header (i.e the offset and
+   *                           the length of the size field)
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: please also document the assumed byte-order

**Comment by rkuhn** [comment]
In principle looks good apart from nitpicks, but needs further fixes to compile successfully.

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala (hunk 1 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: good catch, thanks

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala (hunk 2 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: Ouch, tkanks

File: akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala (hunk 3 above)
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: Wilco

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Framing.scala]: (added above)

**Comment by ktoso** [comment]
Addressed comments and grepped for old `io` usages

**Comment by ktoso** [comment]
Oh c'mon, forgot to fix another thing... Testing locally now, should be all now though...

**Comment by rkuhn** [comment]
LGTM

File: akka-docs/rst/java/stream/migration-guide-2.0-2.4-java.rst [hunk 4]
```diff
@@ -185,3 +185,39 @@ The old behaviour can be achieved by explicitly draining the entity:
 
    response.entity().getDataBytes().runWith(Sink.ignore())
 
+
```
Comment (drewhk) [inline akka-docs/rst/java/stream/migration-guide-2.0-2.4-java.rst]: oh noes, you added to the bottom. We will conflict after merge...

**Comment by drewhk** [comment]
LGTM

**Comment by ktoso** [comment]
Uff, validated finally; Don't mind if I merge then

**Pull request merged**